Analysis/bin/ShuffleMergeSpectral.cxx

```diff
@@ -0,0 +1,773 @@
+               const std::set<TauType>& _tautypes) :
+        name(_name), group_name(_group_name), file_names(_file_names),
+        disabled_branches(_disabled_branches), entry_begin_rel(_begin_rel), entry_end_rel(_end_rel),
+        tau_types(_tautypes), current_n_processed(0), files_n_total(_file_names.size()),
+        entries_file(std::numeric_limits<size_t>::infinity()), total_n_processed(0), total_n_input(0)
+    {
+        if(file_names.empty())
+          throw analysis::exception("Empty list of files for the source '%1%'.") % name;
+        if(!files_n_total)
+          throw analysis::exception("Empty source '%1%'.") % name;
+        dataset_hash = std::hash<std::string>{}(name) % std::numeric_limits<int>::max();
+        datagroup_hash = std::hash<std::string>{}(group_name) % std::numeric_limits<int>::max();
+    }
+
+    SourceDesc(const SourceDesc&) = delete;
+    SourceDesc& operator=(const SourceDesc&) = delete;
+
+    bool DoNextStep()
+    {
+      do {
+        if(current_file_index == files_n_total-1 && current_n_processed >= entries_file) return false;
+        while(!current_file_index || current_n_processed == entries_file) {
+            if(!current_file_index)
+                current_file_index = 0;
+            else
+                ++(*current_file_index);
+            if(*current_file_index >= file_names.size())
+                throw analysis::exception("The expected number of events = %1% is bigger than the actual number of"
+                                          " events in source '%2%'.") % entries_file % name;
+            const std::string& file_name = file_names.at(*current_file_index);
+            current_tuple.reset();
+            current_file = root_ext::OpenRootFile(file_name);
+            current_tuple = std::make_shared<TauTuple>("taus", current_file.get(), true, disabled_branches);
+            entries_file = current_tuple->GetEntries();
+            current_n_processed =  std::floor(entry_begin_rel * entries_file);
+            entries_end = std::floor(entry_end_rel * entries_file);
+            if(!entries_file)
+              throw analysis::exception("Root file inside %1% is empty.") % name;
+        }
+        current_tuple->GetEntry(current_n_processed++);
+        const auto gen_match = static_cast<analysis::GenLeptonMatch>((*current_tuple)().lepton_gen_match);
+        const auto sample_type = static_cast<analysis::SampleType>((*current_tuple)().sampleType);
+        current_tau_type = analysis::GenMatchToTauType(gen_match, sample_type);
+        ++total_n_input;
+      }
+      while (tau_types.find(current_tau_type) == tau_types.end());
+      ++total_n_processed;
+      (*current_tuple)().tauType = static_cast<Int_t>(current_tau_type);
+      (*current_tuple)().dataset_id = dataset_hash;
+      (*current_tuple)().dataset_group_id = datagroup_hash;
+      return true;
+    }
```

Comment:
@kandrosov I found out that with this approach at some point of iteration virtual memory usage exceeds 400Gb and this jobs is killed by task manager. Is this some bug from my side or opening all datasets in parallel can potentially exceed this limit?

Reply:
The RAM exceeds 14 GB...

Reply:
How many dataset groups do you have? Having about 10-20 groups (i.e. 10-20 simultaneously opened files) should not require that much memory.

Reply:
I have 7 data groups, but the concept of the code was to have all data sets opened in parallel, that is ~87 datasets

Reply:
hm... no, this concept should be changed. Only one file from each dataset group should be opened at the time...

Reply:
I see.... in this case everything should be re-done conceptually, I thought in the beginning we agreed about such concept

Reply:
In case of "one file from datagroup opened" how do we mix them within a single data group they just will be sequentially ordered.

Reply:
Hm... in the old implementation of ShuffleMerge only one file from one bin of the given dataset group was opened at the time. My apologies, I probably have missed the point that you wanted to change that concept... 
I remember we have discussed that one of the main limitations of having binned dataset, is the necessity of having opened multiple files per dataset group. Switching to the spectral shuffle&merge was a solution for this issue (among the other issues as well) - with it, dataset group is read consecutively (as in the past) supposing that all datasets in it having equivalent properties, but this time only 1 file per dataset group must be opened at the time.

Reply:
> In case of "one file from datagroup opened" how do we mix them within a single data group they just will be sequentially ordered.

If all datasets must be mixed, then "dataset group" abstraction does not make much sense - the only benefit I see from combining datasets into groups is to reduce the number of sources that must be mixed...

Reply:
> > In case of "one file from datagroup opened" how do we mix them within a single data group they just will be sequentially ordered.
> 
> If all datasets must be mixed, then "dataset group" abstraction does not make much sense - the only benefit I see from combining datasets into groups is to reduce the number of sources that must be mixed...

yes, my bad.... I also had so many doubts about the concept of data group and decided in the end that it was just a logical separation, can you please confirm that now my understanding of the concept is right:
1. We are randomly the order of the files within one data group.
2. We have SourceDesc per every data group, and we are "walking" within shuffled files inside it sequentially.
3. We ensure homogenous occurrence of data group and hope that it also will be relatively homogenous over datasets.
4. Now, it also becomes more clear that this parallelization that we implemented with start and end ratio within a single file actually add additional randomization for dataset occurrence, so, it should be effective not only for saving time....

Reply:
Yes, I confirm that it is also my current understanding. Small remark on point 3. If we group samples with the same physics signature, the requirement of homogeneity of samples from this group across the output dataset is not strictly necessary. Therefore, if we are careful in the definition of the groups, we don't need to "hope that it also will be relatively homogenous over datasets".